File: Python/0232. Implement Queue using Stacks.py
class MyQueue(object):

    def __init__(self):
        """
        in主要負責push, out主要負責pop
        """
        self.stack_in = []
        self.stack_out = []


    def push(self, x):
        """
        有新element進來，就往in裡面push
        """
        self.stack_in.append(x)


    def pop(self):
        """
        Removes the element from in front of queue and returns that elements
        """
        if self.empty():
            return None

        if self.stack_out:
            return self.stack_out.pop()
        else:
            for i in range(len(self.stack_in)):
                self.stack_out.append(self.stack_in.pop())
            return self.stack_out.pop()



    def peek(self):
        """
        Get the front element
        """
        ans = self.pop()
        self.stack_out.append(ans)
        return ans


    def empty(self):
        """
        只要in/out有element，說明隊列不為空
        """
        return not(self.stack_in or self.stack_out)



# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: Java/0232. Implement Queue using Stacks.java
class MyQueue {


    public MyQueue() {
    }
    private Stack<Integer> stackIn = new Stack<>();
    private Stack<Integer> stackOut = new Stack<>();

    public void push(int x) {
        stackIn.push(x);
    }

    public int pop() {
        dumpstackIn();
        return stackOut.pop();
    }

    public int peek() {
        dumpstackIn();
        return stackOut.peek();
    }

    // 如果 stackOut is empty，將 stackIn 中的所有 elements 放到 stackOut 中 -> 準備被 pop 出去
    public boolean empty() {
        return stackIn.isEmpty() && stackOut.isEmpty();
    }

    private void dumpstackIn() {
        if (!stackOut.isEmpty()) return;
        while (!stackIn.isEmpty()) {
            stackOut.push(stackIn.pop());
        }
    }
}


/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */